File: download_mem.php

<?php

include("progress.class.php");

$tasks = 4096; // total tasks to run

$progress = new Progress("Downloading", $tasks);



$port = "/dev/ttyUSB0";
$cmd = "1SendEram";
$eRam = "";
$memfile = fopen("download.mem", "w\n");

$c = exec('stty -F '.$port.' cs8 -parenb -cstopb -echo raw speed 19200');

if(!file_exists($port)) {
	echo "Device doesn't exist.";
	die();
} else {
	$f = fopen($port, "w+");
	fwrite($f, $cmd."\r");
	sleep(0.1);
	$indata="";
	$timediff =0;
	$starttime = date('s');
	$substring = "";

	while(substr($indata,0,1) != "C"){	//look for starting C of Complete
$progress->increase();
		
		$indata = fgets($f);		
	
		if($indata != ""){
		//	echo "INDATA " . "$indata\r";	
			fwrite($memfile, $indata);	#write to download.mem
			fwrite($f,"OK\r");	
			$starttime = time(); //reset timer		
		}
		
		if($indata == "TimeOut"){die("Timeout from controller");}		
		if(substr($indata,0,1) == "C"){
			fclose($memfile);
			$indata = fgets($f); //read final +SENDE and clear from buffer
			exit();
		}		

		$substring = substr($indata,0,1);		
		if($substring =="-"){    		
			print "We encountered an error when receiving $indata". "Please close window and try again\r\n";	
			fclose($memfile);
			exit();		
		}
	
		$endtime = date('s');		
		$timediff = $endtime - $starttime;
	
		if($timediff > '4'){
			print"Serial Timeout! There was no response from the RC210. Please close this window, check your serial connections and try again\r\n";			
			fclose($memfile);
			exit();
		}
		$indata ="";
	}//while
}
$progress->end();
?>

File: download_mem.py
#!/usr/bin/python

import serial, time, sys, fileinput
 
#open and configure serial port
ser = serial.Serial(
    port='/dev/ttyUSB0',
    baudrate=19200,
    parity=serial.PARITY_NONE,
    stopbits=serial.STOPBITS_ONE,
    bytesize=serial.EIGHTBITS,
    timeout = .1	
)

#first, clear out 210 buffer
count = 0
while (count < 3):
 count +=1  
 ser.write("\r".encode())

 time.sleep(.1)
 

#open file for download
file = open("download.mem", "w+")

#send command to rc210 to start download
ser.write("1SendEram\r\n".encode())

indata ="" 
Counter = 0
progresscounter = 0
var = 1

while var == 1:
  inser = str(ser.readline())
  indata = inser.strip
  print (indata)
  if indata == "Complete": #check for first character of "Complete" and exit loop
    break

    Counter = Counter + 1	
  else:
    Counter = 0 
    ser.write("\r".encode())
    file.write(indata)
    #LineCount -= 1 
    progresscounter += 1
    progress = progresscounter / 44
    if( progress > 100 ) : progress = 100
    ser.write("OK\r\n".encode())
	
	
    #print( '\rDownloading: %s (%d%%)' % ("|"*(progress/2), progress)),
    sys.stdout.flush()
 

  
  if Counter > 10 :
    file.close()
    sys.exit("RC210 did not respond. Exiting")

	
print ("\nDownload Complete")
  
  
file.close()

#now for RTC


sys.exit()


